pkg/microservice/aslan/core/renderset/service/renderset.go

```diff
@@ -0,0 +1,319 @@
+	}
+	return ret, nil
+}
+
+func generateValuesYaml(service *commonmodels.Service, args *RendersetCreateArgs, log *zap.SugaredLogger) (string, error) {
+	if args.YamlSource == setting.ValuesYamlSourceFreeEdit {
+		return args.ValuesYAML, nil
+	} else if args.YamlSource == setting.ValuesYamlSourceDefault {
+		return service.HelmChart.ValuesYaml, nil
+	} else if args.YamlSource == setting.ValuesYamlSourceGitRepo {
+		if args.GitRepoConfig == nil {
+			return "", nil
+		}
+
+		var (
+			allValues      []byte
+			fileContentMap sync.Map
+			wg             sync.WaitGroup
+			err            error
+		)
+
+		for index, filePath := range args.GitRepoConfig.ValuesPaths {
+			wg.Add(1)
+			go func(index int, path string) {
+				defer wg.Done()
+				fileContent, err1 := fsservice.DownloadFileFromSource(
+					&fsservice.DownloadFromSourceArgs{
+						CodehostID: args.GitRepoConfig.CodehostID,
+						Owner:      args.GitRepoConfig.Owner,
+						Repo:       args.GitRepoConfig.Repo,
+						Path:       path,
+						Branch:     args.GitRepoConfig.Branch,
+					})
+				if err1 != nil {
+					err = errors.Errorf("fail to download file from git, err: %s, path: %s, repo: %v", err1.Error(), path, *args.GitRepoConfig)
+					return
+				}
+				fileContentMap.Store(index, fileContent)
+			}(index, filePath)
+		}
+		wg.Wait()
+
+		if err != nil {
+			return "", err
+		}
+
+		for i := 0; i < len(args.GitRepoConfig.ValuesPaths); i++ {
+			if content, ok := fileContentMap.Load(i); ok {
+				allValues, err = yaml2.Merge([][]byte{allValues, content.([]byte)})
+				if err != nil {
+					return "", errors.Errorf("fail to merge file, path: %s, repo: %v", args.GitRepoConfig.ValuesPaths[i], *args.GitRepoConfig)
+				}
+			}
+		}
+		return string(allValues), nil
+	}
+	return "", nil
+}
+
+func CreateOrUpdateChartValues(productName, envName string, args *RendersetCreateArgs, userName, requestID string, log *zap.SugaredLogger) error {
+
+	serviceName := args.ServiceName
+
+	serviceOpt := &commonrepo.ServiceFindOption{
+		ProductName: productName,
+		ServiceName: serviceName,
+		Revision:    0,
+	}
+	serviceObj, err := commonrepo.NewServiceColl().Find(serviceOpt)
+	if err != nil {
+		return e.ErrCreateRenderSet.AddDesc(err.Error())
+	}
+	if serviceObj == nil {
+		return e.ErrCreateRenderSet.AddDesc("service not found")
+	}
+	if serviceObj.Type != setting.HelmDeployType {
+		return e.ErrCreateRenderSet.AddDesc("invalid service type")
+	}
+	if serviceObj.HelmChart == nil {
+		return e.ErrCreateRenderSet.AddDesc("missing helm chart info")
+	}
+
+	yamlContent, err := generateValuesYaml(serviceObj, args, log)
+	if err != nil {
+		return e.ErrCreateRenderSet.AddDesc(err.Error())
+	}
+
+	if yamlContent == "" {
+		return e.ErrCreateRenderSet.AddDesc("empty yaml content")
+	}
+	// validate yaml content
+	tMap := map[string]interface{}{}
+	if err = yaml.Unmarshal([]byte(yamlContent), &tMap); err != nil {
+		if err != nil {
+			return e.ErrCreateRenderSet.AddDesc("yaml content illegal")
+		}
+	}
```

Comment:
if it is not a valid yaml, Merge() will fail, so you don't need to validate it again